36 get admininstitutionsid

ValterAro opened the pull request:
Updated OpenAPI specification in the administrative.yml file with the GET operation of admin/institutions/{id} path.
#36 issue

martinkearn commented on .github/workflows/ci-pullrequest-main.yml:
```diff
@@ -146,4 +146,4 @@ jobs:
     
     - name: Docker Build
-      run: docker image build -f src/CentOps.Api/Dockerfile ./src
+      run: docker image build -f src/CentOps.Api/Dockerfile ./src
```
This does not seem to be a deliberate change?

yahya130 replied:
Accidental, I will leave remove this from the PR

martinkearn commented on src/CentOps.Api/Controllers/PublicParticipantController.cs:
```diff
@@ -45,5 +47,26 @@ public async Task<ActionResult<ParticipantResponseModel>> Get(string id)
+            try
+            {
+                var response = await _store.UpdateState(id, newState).ConfigureAwait(false);
+                return Ok(_mapper.Map<ParticipantStateReponseModel>(response));
```
super nit: Just from a readability perspective, I always favour having a var for the thing you are returning separate from the return statement itself. I find it easier to read that way, could just be me though, feel free to ignore.
```suggestion
                var result = _mapper.Map<ParticipantStateReponseModel>(response);
                return Ok(result);
```

yahya130 replied:
Agreed just a short cut this, but readability would be a reason to commit this change

martinkearn commented on src/CentOps.Api/Controllers/PublicParticipantController.cs:
```diff
@@ -45,5 +47,26 @@ public async Task<ActionResult<ParticipantResponseModel>> Get(string id)
+            {
+                return NotFound(ex.Message);
+            }
+            catch (Exception ex) when (ex is ModelNotFoundException<ParticipantDto> or ArgumentException or ArgumentNullException)
```
Would `when (ex is ModelNotFoundException<ParticipantDto>` ever get hit because you are checking for that on line 62?

yahya130 replied:
Good spot, will update

martinkearn commented on src/CentOps.Api/Dockerfile:
```diff
@@ -10,14 +10,6 @@ RUN dotnet publish -c Release -o out
 
 # Build runtime image
 FROM mcr.microsoft.com/dotnet/aspnet:6.0
-
```
Just double checking this is deliberate? If it is, please ignore

guypritchard commented on src/CentOps.Api/Controllers/PublicParticipantController.cs:
```diff
@@ -45,5 +47,26 @@ public async Task<ActionResult<ParticipantResponseModel>> Get(string id)
+        {
+            try
+            {
+                var response = await _store.UpdateState(id, newState).ConfigureAwait(false);
```
I think on the public API we need to look at the API key we've been given here too - a participant shouldn't be able to alter the state of another participant just because they have the ID.  Or is this somewhere else in the code I'm not seeing?
If the Id and the Id of the participant doesn't match we shouldn't proceed.  Have a look at the code which adds the participant Id.

Note - this pushes on there always being unique API keys... The current functionality allows duplicates - but as you can probably guess this means the code may misidentify the participant.

yahya130 replied:
I implemented this at first, but I realized and you mentioned on the teams chat too, that this is handled by the authentication logic and would be good to capture any further functionality there

decodingahmed commented on src/CentOps.Api/Program.cs:
```diff
@@ -49,6 +49,7 @@ public static void Main(string[] args)
 
             _ = app.MapControllers();
 
+
```
supernit: remove empty line please
```suggestion
```

decodingahmed commented on src/CentOps.Api/Services/CosmosDbService.cs:
```diff
@@ -411,5 +412,36 @@ async Task<IEnumerable<ParticipantDto>> IParticipantStore.GetAll(IEnumerable<Par
+                throw new ModelNotFoundException<ParticipantDto>(id);
+            }
+
+            var validStateChanges = new List<ParticipantState>() { ParticipantState.Online, ParticipantState.Offline };
```
Could you make this list a static variable at the top of the class please?  This line will recreate this list for every request so lets reduce those unnecessary allocations.

guypritchard commented on src/CentOps.Api/Models/ParticipantResponseModel.cs:
```diff
@@ -17,5 +17,7 @@ public class ParticipantResponseModel
 
         public ParticipantStatus Status { get; set; }
+
+        public ParticipantState State { get; set; }
```
Do we need a new property for this?  Could we not just add the states here to ParticipantStatus or does this truly need to be something else?

decodingahmed commented on src/CentOps.Api/Services/CosmosDbService.cs:
```diff
@@ -411,5 +412,36 @@ async Task<IEnumerable<ParticipantDto>> IParticipantStore.GetAll(IEnumerable<Par
+
+            participant.State = newState;
+
+            var response = await store.Update(participant).ConfigureAwait(false);
```
There is a way to just update a property in Cosmos. I'll call to walk through how we can do that. It will save you from having to fetch then update the whole object.

yahya130 replied:
Agreed, I had a little trouble last minute when the test failed trying to mock up (https://github.com/buerokratt/CentOps/pull/88/commits/b9375e6e8edc80a40b7aa8451a645a489feadc4f#diff-70e24f7e44ff7fe71ab6cd266d2f09e6c034f8d4ce99c4c186067be07e54b59c) the cosmosdb patch method so I reverted back to the update method

decodingahmed commented on src/CentOps.Api/Models/ParticipantState.cs:
```diff
@@ -0,0 +1,9 @@
+﻿namespace CentOps.Api.Models
+{
+    public enum ParticipantState
```
Please correct me if this is wrong, but I think this `ParticipantState` and the existing `ParticipantStatus` enums are trying to do the same thing @guypritchard, what do you think?

guypritchard replied:
I added a comment about this - I think so, I think we could get away with adding these states to the existing enum.

yahya130 replied:
I was not sure whether ParticipantStatus was being used in other areas of the code, but to me having read this doc (https://github.com/buerokratt/CentOps/blob/main/docs/Participants%20States.md), there were some grey areas for example, offline/online vs active/disabled, also feel based off that doc and other areas of the project the naming of 'ParticipantState' over 'ParticipantStatus' seemed more natural, so for me perhaps updating the naming to 'ParticipantState' might be up for consideration, but of course happy with 'ParticipantStatus' if that is the preference :)

yahya130 commented on src/CentOps.Api/Controllers/PublicParticipantController.cs:
```diff
@@ -45,5 +48,31 @@ public async Task<ActionResult<ParticipantResponseModel>> Get(string id)
                 : NotFound(id);
         }
+
+        [HttpPut("my/state")]
```
Just a question, is this the most restful way to do this in terms of the new path? I know the definition of restful is a very open topic, but would be good to understand :) also I noticed the id param is removed, how is the id passed into the request now?

decodingahmed replied:
In terms of RESTful pattern, here's the thinking:
This is the complete path of this endpoint: `/public/participants/my/state`

`/public/participants` - pluralised resource path
`/my` - this is equivalent to having an `id` but we specify "my" so it still fits the standard pattern
`/state` - is allowed as a subpath (aka a micro-endpoint)

Regarding the removed `id` param, at the moment none of the participants know their own ID so asking them to provide it would mean that we would need to make a change elsewhere.

Secondly, we are able to identify them using their API key so when they call CentOps, it pulls their record from the DB to validate that the API key exists. Other information about the participant, like the `id`, is loaded up in the request from within authentication (see `ApiUserClaimsProvider.cs`)

Hope all that made sense. Please ping me if you want to talk more about this dude. Thanks for the comment @yahya130

guypritchard commented on src/CentOps.Api/Services/CosmosDbService.cs:
```diff
@@ -317,7 +317,7 @@ WHERE STARTSWITH(c.pk, 'participant', false)
                         AND c.status != @status",
                 ("@apiKey", apiKey),
-                ("@status", ParticipantStatusDto.Disabled));
+                ("@status", ParticipantStatusDto.Deleted));
```
Thinking out loud:  Should this only return an API key for Active participants?  (e.g. not Delete OR Disabled?)

guypritchard replied:
We think Disabled actually means active and offline.  Keys should still be valid in this situation.

guypritchard replied:
I think we're going to no longer expose partition keys externally.

guypritchard commented on src/CentOps.Api/Services/ApiUserClaimsProvider.cs:
```diff
@@ -21,11 +21,12 @@ public ApiUserClaimsProvider(IParticipantStore participantStore)
                 result = new ApiUser(new[]
                 {
                     new Claim("id", user.Id!),
+                    new Claim("pk", user.PartitionKey!),
```
Is putting the partition key in a claim a bit weird?  Is this to make querying later on easier?

guypritchard replied:
We don't think we'll need this anymore.

decodingahmed replied:
PartitionKey claim has been removed

guypritchard commented on src/CentOps.Api/Services/CosmosDbService.cs:
```diff
@@ -411,5 +411,26 @@ async Task<IEnumerable<ParticipantDto>> IParticipantStore.GetAll(IEnumerable<Par
             return results;
         }
+
+        public async Task<ParticipantDto> UpdateState(string id, string partitionKey, ParticipantStatusDto newStatus)
```
Can we not deduce the partition key here?  Or is there something we can't infer?

decodingahmed replied:
This method no longer takes the `partitionKey` param.

ValterAro committed:
Merge pull request #87 from buerokratt/main

Update from the main branch

ValterAro committed:
Update administrative.yml spec

Added specification for GET operation on the admin/institutions/{id} path. Also added admin/institutions tag.
